Add option to shorten a display name

garethj2 opened the pull request:
We need some support for shortened display names.

This differs from the relative display names that exist in Haxall. In our use case, we may not have the parent when we have a shortened display name.

So how to handle it? Firstly we need to understand that display names can be processed in different ways (`dis`, `disMacro`, `name` etc). A long display name normally incorporates refs to other records. This uses `disMacro`. For example, if there was a `disMacro` tag on a point with `$equipRef $navName`, the result may include the display names for all the parents of the record and the point itself.

An easy to deal with this is by skipping any `Ref` values when processing the `disMacro`. Therefore the result is really `$equipRef`.

I admit this isn't a perfect solution. I'd argue the relative display names in `Haxall` are worse since you always need to know the parent. This is a pain when you want to render short display names from a list and not a tree control.

gretadj2 commented:
> An easy to deal with this is by skipping any Ref values when processing the disMacro. Therefore the result is really $equipRef.

Just to double check if I understood correctly, did you mean that "the result is really `$navName`", right?

I agree the solution is not perfect, because if for example the user decides to add some separation characters in the disMacro, you could end up with a strange result: `$siteRef / $floorRef / $navName` => `/ / a navName`

But at the same time I know it's a complex topic, because different contexts often lead to different display names. For example, we had a few customers that wanted to use the `disKey` functionality only on some points, so we had to check the tags in the Dict to understand which one to prioritize.
So I think this may be an additional tool that could help, even if it's not a panacea.

kianj2 commented:
Interesting point on the separation characters @gretadj2. It makes me wonder what that perfect solution could be later on down the road

garethj2 commented:
I've taken everyone's feedback on board and have redesigned the approach here.

@gretadj2 I think your comments regarding how customers use separators is really important. I think @EliteScientist has raised some good ideas with me too.

I've simplified the approach here. Now if a display name is shortened, it moves the precedence of resolving the `disMacro` further down the list. I've also added support for finding `navName` (again fairly low on precedence). There is no more overly clever processing of the display name. It's a slightly more advanced version of using either `navName` for the short name if you think about it.

Let me know what you think. I originally created this PR to kick around a few ideas and see what works best for people.

gretadj2 commented:
👍  I like it, thanks @garethj2 !

kianj2 commented on spec/core/Util.spec.ts:
```diff
@@ -645,6 +645,25 @@ describe('util', function (): void {
 		})
+
+		it('returns navName', function (): void {
+			expect(dictToDis(new HDict({ dis: 'navName' }))).toBe('navName')
```
Shouldn't this test be using a dict with a `navName` tag instead of a `dis` tag?

kianj2 commented on spec/core/Util.spec.ts:
```diff
@@ -645,6 +645,25 @@ describe('util', function (): void {
+				disMacro: HStr.make('    $siteRef $equipRef $navName   '),
+			})
+
+			// Shortening a display name will remove any refs from output of
```
Another super minor thing, but just to avoid future confusion I think it would be good to remove this comment

kianj2 reviewed (approved):
I think the changes are good! I left some super minor comments based on the changes. Thanks Gareth

garethj2 commented on src/core/util.ts:
```diff
@@ -375,31 +375,44 @@ export interface LocalizedCallback {
  * 3. 'disKey' maps to qname locale key.
  * 4. 'name' tag.
  * 5. 'tag' tag.
+ * 6. 'navName' tag.
```
This was sloppy of me. I shouldn't try and squeeze in commits between meetings :). Thanks.

kianj2 replied:
No worries at all :) it's appreciated you're even trying to squeeze these changes in between meetings!
